Sources/Target/Common/ThreadBase.cpp

```diff
@@ -39,5 +39,36 @@ ErrorCode ThreadBase::beforeResume() {
 
   return kSuccess;
 }
+
+void ThreadBase::setBreakpointStopReason() {
+  auto *hwBpm = process()->hardwareBreakpointManager();
+  if (hwBpm == nullptr) {
+    return;
+  }
+
+  BreakpointManager::Site site;
+  int bpIdx = hwBpm->hit((Target::Thread *)this, site);
+  if (bpIdx >= 0) {
+    _stopInfo.watchpointIndex = bpIdx;
+    _stopInfo.watchpointAddress = site.address;
+
+    switch (static_cast<int>(site.mode)) {
+    case BreakpointManager::kModeExec:
+      _stopInfo.reason = StopInfo::kReasonBreakpoint;
+      break;
+    case BreakpointManager::kModeWrite:
+      _stopInfo.reason = StopInfo::kReasonWriteWatchpoint;
+      break;
+    case BreakpointManager::kModeRead:
+      _stopInfo.reason = StopInfo::kReasonReadWatchpoint;
+      break;
+    case BreakpointManager::kModeRead | BreakpointManager::kModeWrite:
+      _stopInfo.reason = StopInfo::kReasonAccessWatchpoint;
+      break;
+    default:
+      DS2BUG("Invalid mode");
```

Comment:
Nit: We keep log strings with no capital letter at the beginning, so `"invalid mode"` instead of `"Invalid mode"`.